### PR TITLE
Momchil/mode solver

### PR DESCRIPTION
**Pull request by momchil-flex**
Some highlights:

- New `ModeFieldData` and `ScalarModeFieldData` are like `FieldData` but defined with an additional coordinate, `mode_index`.
- Revmoed `freq` from `ModeSolver` and instead `ModeSolver.solve` now takes a list of frequencies. `ModeSolver.to_source` takes a `source_time` instead of an `fwidth`.
- Frontend and backend tests pass, docs notebooks still need to be fixed.

And some thoughts:
- I have a `sel_mode_index` method on `ModeFieldData` right now, used in the backend. Should we instead consider introducing a `sel` method for all `CollectionData` objects, which iterates over all data and calls xarray `sel` if coordinate exists?
- I think we can combine `Tidy3dData` and `MonitorData` into just Tidy3dData? Right now `ModeFieldData` is technically a subclass of `MonitorData`, even though it's not really associated to a montior.
- `CollectionData`-s like `FieldData` have some nice properties, so I don't think we can convert them to Dataset, but maybe everything that's not a `CollectionData` can be converted to an xarray `DataArray` as per #228 .

### CODE REVIEW

**Comment by tylerflex** [comment]
> I have a sel_mode_index method on ModeFieldData right now, used in the backend. Should we instead consider introducing a sel method for all CollectionData objects, which iterates over all data and calls xarray sel if coordinate exists?

That sounds useful, but maybe not necessary yet?

> I think we can combine Tidy3dData and MonitorData into just Tidy3dData? Right now ModeFieldData is technically a subclass of MonitorData, even though it's not really associated to a montior.

Let's just make ModeFieldData a subclass of Tidy3DData but leave the other stuff how it is.  It's set up this way to be able to independently configure MonitorData and CollectionData.  In general (and related to the point above) we should consider a refactor of all of this stuff.

> CollectionData-s like FieldData have some nice properties, so I don't think we can convert them to Dataset, but maybe everything that's not a CollectionData can be converted to an xarray DataArray as per https://github.com/flexcompute/tidy3d/issues/228 .

Yea that would be ideal.  I think for each "monitor data"-like object we can:
* Subclass `XR.DataArray`
* Hard code the `dims` specific to each type of data.

And for each "collection data" like data, similar to now:
* data_dict mapping the field name to the corresponding DataArray.
* get_item and @property automatically added for each field.
* Hard code the `dims` common to each monitor data in the collection.

Then it would cut down on the complexity of everything quite a bit.

File: tidy3d/components/monitor.py [hunk 1]
```diff
@@ -382,5 +386,32 @@ def _dir_arrow(self) -> Tuple[float, float, float]:
...
+    """
+
+    _data_type: Literal["ModeSolverData"] = pydantic.Field("ModeSolverData")
+
+    def storage_size(self, num_cells: int, tmesh: int) -> int:
+        # fields store 6 complex numbers per grid cell, per frequency, per mode.
+        field_size = 6 * BYTES_COMPLEX * num_cells * len(self.freqs) * self.mode_spec.num_modes
+        # effective index stores 1 complex number per frequency per mode
+        neff_size = BYTES_COMPLEX * len(self.freqs) * self.mode_spec.num_modes
+        return field_size * neff_size
```
Comment (tylerflex) [inline tidy3d/components/monitor.py:411]: I'm confused, shouldn't this be +?

File: tidy3d/plugins/mode/mode_solver.py [hunk 2]
```diff
@@ -1,121 +1,83 @@
...
+        ..., title="Simulation", description="Simulation defining all structures and mediums."
+    )
 
+    plane: Box = pydantic.Field(
+        ..., title="Plane", description="Cross-sectional plane in which the mode will be computed."
+    )
 
-class ModeSolver:
-    """Interface for creating ``Mode`` objects."""
+    @pydantic.validator("plane", allow_reuse=True, always=True)
```
Comment (tylerflex) [inline tidy3d/plugins/mode/mode_solver.py:41]: dont really need allow_reuse here.

File: tidy3d/plugins/mode/mode_solver.py [hunk 3]
```diff
@@ -125,110 +87,148 @@ def solve(self, mode_spec: ModeSpec) -> List[ModeInfo]:
...
+        eps_xx = self.simulation.epsilon(plane, "Ex", freq)
+        eps_yy = self.simulation.epsilon(plane, "Ey", freq)
+        eps_zz = self.simulation.epsilon(plane, "Ez", freq)
 
-            modes.append(mode_info)
+        return np.stack((eps_xx, eps_yy, eps_zz), axis=0)
 
-        return modes
+    def solver_eps(self, freq: float) -> Array[complex]:
+        """Get the diagonal permittivity as supplied to the sovler, with the nomral axis rotated to
```
Comment (tylerflex) [inline tidy3d/plugins/mode/mode_solver.py:153]: normal (typo)

File: tidy3d/plugins/mode/mode_solver.py [hunk 4]
```diff
@@ -125,110 +87,148 @@ def solve(self, mode_spec: ModeSpec) -> List[ModeInfo]:
...
-        eps_zz = self.simulation.epsilon(plane, "Ez", self.freq)
+        # swap axes to plane coordinates (normal_axis goes to z)
+        eps_sim_ax = (eps_xx, eps_yy, eps_zz)
+        eps_zz, (eps_xx, eps_yy) = self.plane.pop_axis(eps_sim_ax, axis=self.normal_axis)
 
+        # construct eps to feed to mode solver
         return np.stack((eps_xx, eps_yy, eps_zz), axis=0)
 
+    def rotate_field_coords(self, field):
+        """move the propagation axis=z to the proper order in the array"""
```
Comment (tylerflex) [inline tidy3d/plugins/mode/mode_solver.py:172]: capitalize Move and add period

File: tidy3d/plugins/mode/mode_solver.py [hunk 5]
```diff
@@ -125,110 +87,148 @@ def solve(self, mode_spec: ModeSpec) -> List[ModeInfo]:
...
     ) -> ModeSource:
         """Creates ``ModeSource`` from a Mode + additional specifications.
 
         Parameters
         ----------
         mode_spec : ModeSpec
             :class:`ModeSpec` object containing specifications of mode.
-        fwidth : float
-            Standard deviation of ``GaussianPulse`` of source (Hz).
+        source_time: SourceTime
```
Comment (tylerflex) [inline tidy3d/plugins/mode/mode_solver.py:230]: probably should use `:class:` stuff in here for source time

**Review by tylerflex — CHANGES_REQUESTED** [review]
Looks good.  Just a few cosmetic things with the docstrings and a question about storage size for mode fields + neff data.

**Comment by tylerflex** [comment]
Feel free to merge once clearing up my comments.

File: tidy3d/plugins/mode/mode_solver.py (hunk 2 above)
Comment (tylerflex) [reply tidy3d/plugins/mode/mode_solver.py:41]: Also, can you change the docstring?  `Mode` -> `:class:ModeSpec`?

**Comment by momchil-flex** [comment]
Nice, thanks for catching all of those.

File: tidy3d/plugins/mode/mode_solver.py (hunk 5 above)
Comment (momchil-flex) [reply tidy3d/plugins/mode/mode_solver.py:230]: So same for ModeSpec in `mode_spec : ModeSpec` then?

**Pull request merged**